file/writable_file_writer.cc

```diff
@@ -28,6 +28,11 @@ IOStatus WritableFileWriter::Create(const std::shared_ptr<FileSystem>& fs,
                                     const FileOptions& file_opts,
                                     std::unique_ptr<WritableFileWriter>* writer,
                                     IODebugContext* dbg) {
+  if (file_opts.use_direct_writes &&
+      0 == file_opts.writable_file_max_buffer_size) {
+    return IOStatus::InvalidArgument(
+        "Direct write requires writable_file_max_buffer_size > 0");
+  }
```

Comment:
Should this be in PosixFileSystem::OpenWritableFile (and possibly for Windows as well) instead of here?

Reply:
Alternatively, this should could create the file and see if use_direct_io() is true and then return an error.  Obviously that might be a little less efficient but would be more accurate.

Reply:
WritableFileWriter is one layer above PosixFile and Windows file, and `file_opts.writable_file_max_buffer_size` is used to initialize `WritableFileWriter::max_buffer_size_`. Therefore, we can and should add this check in `WritableFileWriter`.

I think detecting and returning early on the illegal combination is better, especially the illegal combination is **not** the default value. It indicates the application deliberately sets the options values so, and RocksDB should warn the application early.

Reply:
But there are other FileSystems besides Windows and Posix to consider -- HDFS, ZenFS, S3, etc.  Are you suggesting these will all behave the same way or the restriction should be universal?

If a FileSystem does not support direct IO, isn't the direct IO option ignored but this check would cause a bad Status.  There should probably be a FileSystem::SupportsDirectIO(const std::string& fname) const method to test whether or not a FileSystem can do Direct IO on a given file (there are already tests that try to implement such a call).  IMO this would be better than assuming that the FileSystem will pay attention to these options and cause problems later for this WritableFileWriter.

Reply:
Just to clarify, the infinite loop occurs in `WritableFileWriter`, not `<XX>WritableFile`, although it does require the file system to support direct IO.

> If a FileSystem does not support direct IO, isn't the direct IO option ignored but this check would cause a bad Status.

If file system does not support direct IO, then `DB::Open()` will have already failed. But RocksDB may internally use `WritableFileWriter` bypassing the db open, thus we cannot rely on the checks in db open.

I do not consider it a bad status. The combination itself does not make much sense: user demands direct IO, but deliberately requires that RocksDB allocates no memory buffer. I think it is better to return an error so that user knows exactly what goes wrong. If I were a user, I prefer learning about this as early as possible, instead of letting RocksDB (silently) determine that direct IO is not supported on this specific platform. Whenever we see this combination, we should return an error.

Reply:
Imagine a RocksDB developer writing some internal code using `WritableFileWriter`. We are setting up this situation so that the developer cannot rely on the validation in db open.

We return error only if all the following conditions meet:
1. FS supports direct IO
2. user sets use_directIO to true
3. user sets buffer_size to 0

If we do not return error early (when 2 and 3 meet), then it is possible that the file system does not support direct IO (1 does not meet), users sets the illegal combination of options (2 and 3) and his program works. Later, when he continues his work on another platform with direct IO support, he will find that his code no longer works.

Reply:
I do not see the place in the code that checks that DB::Open will fail if direct_writes is not supported.  The code in DB::Open that checks that files can be opened for direct read but I do not see any code that checks for direct write.  And I am not sure what prevents a "use_direct_io_for_flush_and_compaction=false" while setting "FileOptions::use_direct_writes=true".  It is unclear to me how we know that the FileSystem will support direct writes, or how we know it supports it for a given file.

Reply:
> checks that files can be opened for direct read but I do not see any code that checks for direct write

I am not aware of a file system that supports direct read but not write, and I doubt any FS/Env/EnvWrapper will do that.

As the PR description explains, it's not important whether the file system supports direct IO or not. As long as we see a combination of `use_direct_io_for_flush_and_compaction=true` and `writable_file_max_buffer_size=0`, we should inform the user that this combination is invalid. This is the reason why I insist on performing the check before checking for support for direct IO. If you have strong opinion against this, can you use an example to help me understand the concern?

> what prevents a "use_direct_io_for_flush_and_compaction=false" while setting "FileOptions::use_direct_writes=true"

If the WritableFileWriter and WritableFile are created by `DB`, then FileOptions.use_direct_writes will be set using DBOptions.use_direct_io_for_flush_and_compaction. Please refer to `DBImpl::DBImpl()`. You can always create a WritableFileWriter and WritableFile without involving the DB. In both cases, we perform the check in `WritableFileWriter` because the `writable_file_max_buffer_size` is a setting for `WritableFileWriter`.